### PR TITLE
Fix audit concerns

### PR DESCRIPTION
**Pull request by luzzif**
This PR addresses audit concerns.

### CODE REVIEW

File: contracts/ERC20StakingRewardsDistribution.sol [hunk 1]
```diff
@@ -105,15 +107,25 @@ contract ERC20StakingRewardsDistribution {
...
+        require(_rewardTokenAddresses.length <= 5, "SRD27");
 
-        secondsDuration = _endingTimestamp - _startingTimestamp;
         // Initializing reward tokens and amounts
-        for (uint32 _i = 0; _i < _rewardTokenAddresses.length; _i++) {
+        for (uint8 _i = 0; _i < _rewardTokenAddresses.length; _i++) {
             address _rewardTokenAddress = _rewardTokenAddresses[_i];
+
+            // checking for duplicates
+            for (uint8 _j = 0; _j < _rewardTokenAddresses.length; _j++) {
```
Comment (jellegerbrandy) [inline contracts/ERC20StakingRewardsDistribution.sol]: you can save some gas by chaing this line to:

`for (uint8 _j = i+1; _j < _rewardTokenAddresses.length; _j++)`

File: test/erc20-distribution/multi-rewards-single-stakable/initialization.test.js [hunk 2]
```diff
@@ -62,9 +66,58 @@ describe("ERC20StakingRewardsDistribution - Multi rewards, single stakable token
         }
     });
 
+    it("should fail when more than 10 reward tokens are specified", async () => {
```
Comment (ben-kaufman) [inline test/erc20-distribution/multi-rewards-single-stakable/initialization.test.js]:
```suggestion
    it("should fail when more than 5 reward tokens are specified", async () => {
```

File: contracts/ERC20StakingRewardsDistribution.sol (hunk 1 above)
Comment (luzzif) [reply contracts/ERC20StakingRewardsDistribution.sol]: Fixed, thanks for the suggestion.

File: test/erc20-distribution/multi-rewards-single-stakable/initialization.test.js (hunk 2 above)
Comment (luzzif) [reply test/erc20-distribution/multi-rewards-single-stakable/initialization.test.js]: Oops, fixed this. Thanks.

File: contracts/ERC20StakingRewardsDistribution.sol [hunk 3]
```diff
@@ -186,16 +220,47 @@ contract ERC20StakingRewardsDistribution {
...
+        onlyOwner
+        onlyStarted
+    {
+        consolidateReward();
+        uint256[] memory _recoveredUnassignedRewards =
+            new uint256[](rewards.length);
+        for (uint256 _i; _i < rewards.length; _i++) {
+            Reward storage _reward = rewards[_i];
+            address _rewardToken = _reward.token;
+            if (_token == _rewardToken) {
```
Comment (jellegerbrandy) [inline contracts/ERC20StakingRewardsDistribution.sol:247]: So if you change this line to 

`if (_token == address(0) || _token == _rewardToken) {`

you can have a single function definition for recovering all rewards together, or a specific one,  and do not need to  duplicate all the code from `recoverRewards` in a new function `recoverSpecificUnassignedRewards`

File: contracts/ERC20StakingRewardsDistribution.sol [hunk 4]
```diff
@@ -172,11 +184,33 @@ contract ERC20StakingRewardsDistribution {
         emit Canceled();
     }
 
-    function recoverUnassignedRewards() external onlyStarted {
+    function recoverRewardAfterCancel(address _token)
```
Comment (jellegerbrandy) [inline contracts/ERC20StakingRewardsDistribution.sol:190]: Not sure why this was added?

File: contracts/ERC20StakingRewardsDistribution.sol (hunk 3 above)
Comment (jellegerbrandy) [reply contracts/ERC20StakingRewardsDistribution.sol:247]: Same holds for claimAllSpecific

File: contracts/ERC20StakingRewardsDistribution.sol [hunk 5]
```diff
@@ -186,16 +228,48 @@ contract ERC20StakingRewardsDistribution {
...
-            if (!_atLeastOneNonZeroRecovery && _recoverableRewards > 0)
-                _atLeastOneNonZeroRecovery = true;
-            _recoveredUnassignedRewards[_i] = _recoverableRewards;
-            IERC20(_reward.token).safeTransfer(owner, _recoverableRewards);
+            if (_recoverableRewards > 0) {
+                _recoveredUnassignedRewards[_i] = _recoverableRewards;
+                IERC20(_reward.token).safeTransfer(owner, _recoverableRewards);
+            }
         }
-        require(_atLeastOneNonZeroRecovery, "SRD22");
```
Comment (ben-kaufman) [inline contracts/ERC20StakingRewardsDistribution.sol:194]: Could also remove the error code from the top list, unless it's annoying with renumbering...

File: contracts/ERC20StakingRewardsDistribution.sol [hunk 6]
```diff
@@ -158,25 +170,55 @@ contract ERC20StakingRewardsDistribution {
...
 
-    function cancel() external onlyOwner {
+    function cancel() external override onlyOwner {
         require(initialized && !canceled, "SRD19");
         require(block.timestamp < startingTimestamp, "SRD08");
+        canceled = true;
+        emit Canceled();
+    }
+
+    function recoverRewardsAfterCancel(address _token)
```
Comment (ben-kaufman) [inline contracts/ERC20StakingRewardsDistribution.sol]: Shouldn't have a `_token` here, it's unused

File: contracts/ERC20StakingRewardsDistribution.sol [hunk 7]
```diff
@@ -158,25 +170,55 @@ contract ERC20StakingRewardsDistribution {
...
+        external
+        override
+        onlyOwner
+    {
+        require(canceled, "SRD29");
         for (uint256 _i; _i < rewards.length; _i++) {
             Reward storage _reward = rewards[_i];
             IERC20(_reward.token).safeTransfer(
                 owner,
                 IERC20(_reward.token).balanceOf(address(this))
```
Comment (ben-kaufman) [inline contracts/ERC20StakingRewardsDistribution.sol:185]: Nit but could check here if balance is 0 before sending it (and save the result to use in the send).

File: contracts/ERC20StakingRewardsDistribution.sol (hunk 5 above)
Comment (luzzif) [reply contracts/ERC20StakingRewardsDistribution.sol:194]: Solved in a610a9c663acfcbcf36d9b0dde5136d2ade18599

File: contracts/ERC20StakingRewardsDistribution.sol (hunk 6 above)
Comment (luzzif) [reply contracts/ERC20StakingRewardsDistribution.sol]: Solved in 988c11d7179a8d3fef4b6f5421e6945236c5e5cc

File: contracts/ERC20StakingRewardsDistribution.sol (hunk 7 above)
Comment (luzzif) [reply contracts/ERC20StakingRewardsDistribution.sol:185]: Solved in 08729725853b070fad676d2a74616c8a77673a4a

**Pull request merged**